src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs

```diff
@@ -1,10 +1,18 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+
 namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
 {
-    // This is basically just a marker type now so that HubProtocol can return a common base class other than object.
     public abstract class HubMessage
     {
+        public IDictionary<string, string> Headers { get; set; }
+
+        protected HubMessage()
+        {
+            Headers = new Dictionary<string, string>();
```

Comment:
Can we make this lazy

Reply:
Do we care about multi-threaded access? If not, then sure. If so, I'm not sure it's actually going to save anything.